Examples/Algorithms/TrackFinding/src/SpacePointMaker.cpp

```diff
@@ -0,0 +1,153 @@
+    if (ref.volume() != cmp.volume()) {
+      return false;
+    }
+    // within the same volume hierarchy only consider layers
+    return (ref.layer() == cmp.layer());
+  };
+  auto geoSelBeg = m_cfg.geometrySelection.begin();
+  auto geoSelEnd = m_cfg.geometrySelection.end();
+  // sort geometry selection so the unique filtering works
+  std::sort(geoSelBeg, geoSelEnd);
+  auto geoSelLastUnique = std::unique(geoSelBeg, geoSelEnd, isDuplicate);
+  if (geoSelLastUnique != geoSelEnd) {
+    ACTS_WARNING("Removed " << std::distance(geoSelLastUnique, geoSelEnd)
+                            << " geometry selection duplicates");
+    m_cfg.geometrySelection.erase(geoSelLastUnique, geoSelEnd);
+  }
+  ACTS_INFO("Space point geometry selection:");
+  for (const auto& geoId : m_cfg.geometrySelection) {
+    ACTS_INFO("  " << geoId);
+  }
+}
+
+ActsExamples::ProcessCode ActsExamples::SpacePointMaker::execute(
+    const AlgorithmContext& ctx) const {
+  const auto& sourceLinks =
+      ctx.eventStore.get<IndexSourceLinkContainer>(m_cfg.inputSourceLinks);
+  const auto& measurements =
+      ctx.eventStore.get<MeasurementContainer>(m_cfg.inputMeasurements);
+
+  SimSpacePointContainer spacePoints;
+  spacePoints.reserve(sourceLinks.size());
+
+  for (Acts::GeometryIdentifier geoId : m_cfg.geometrySelection) {
+    // select volume/layer depending on what is set in the geometry id
+    auto range = selectLowestNonZeroGeometryObject(sourceLinks, geoId);
+    // groupByModule only works with geometry containers, not with an
+    // arbitrary range. do the equivalent grouping manually
+    auto groupedByModule = makeGroupBy(range, detail::GeometryIdGetter());
+
+    for (auto [moduleGeoId, moduleSourceLinks] : groupedByModule) {
+      // find corresponding surface
+      const Acts::Surface* surface =
+          m_cfg.trackingGeometry->findSurface(moduleGeoId);
+      if (not surface) {
+        ACTS_ERROR("Could not find surface " << moduleGeoId);
+        return ProcessCode::ABORT;
+      }
+
+      for (auto sourceLink : moduleSourceLinks) {
+        // extract a local position/covariance independent from the concrecte
+        // measurement content. since we do not know if and where the local
+        // parameters are contained in the measurement parameters vector, they
+        // are transformed to the bound space where we do know their location.
+        // if the local parameters are not measured, this results in a
+        // zero location, which is a reasonable default fall-back.
+        auto [localPos, localCov] = std::visit(
+            [](const auto& meas) {
+              auto expander = meas.expander();
+              Acts::BoundVector par = expander * meas.parameters();
+              Acts::BoundSymMatrix cov =
+                  expander * meas.covariance() * expander.transpose();
+              // extract local position
+              Acts::Vector2 lpar(par[Acts::eBoundLoc0], par[Acts::eBoundLoc1]);
```

Comment:
Do we assume the measurement is 2-dimensional here? In the case of strip detectors, we still have to use another algorithm to form the space point using two 1-dimensional measurements  on the module, right?

Reply:
There is no such assumption as explained in the code comment above the `std::visit` call and in the algorithm documentation. A missing loc{0,1} will just lead to a zero value in that coordinate; if both are missing, the space point position will be the module center position.

This is a test framework, so if you want to do stupid things, I will let you do stupid things. I can also imagine a setup with tiny detector module sand sparse events, such that you have at most one particle per module per event. In that case, it would even make sense to generate space points without local positions.

Reply:
And yes, for strips where a space-point originates from two 1d measurements, we need a different algorithm altogether. With these changes, the seeding algorithm already supports taking multiple space point collections as input, so we would just need to add another space point algorithm without any further modifications.

Reply:
I see. In principle, the space point only needs to take a global position and its covariance, which could be taken from either a  1D or 2D or even empty measurement. Thank you for the explanations.